tests/interfaces/test_batch_jax.py

```diff
@@ -610,3 +614,176 @@ def cost_fn(a, p, device):
             ]
         )
         assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_independent_expval(self, execute_kwargs, interface):
+        """Tests computing an expectation value that is independent trainable
+        parameters."""
+        dev = qml.device("default.qubit", wires=2)
+        params = jnp.array([0.1, 0.2, 0.3])
+
+        def cost(a, cache):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.RY(a[2], wires=0)
+                qml.expval(qml.PauliZ(1))
+
+            res = qml.interfaces.batch.execute(
+                [tape], dev, cache=cache, interface=interface, **execute_kwargs
+            )
+            return res[0][0]
+
+        res = jax.grad(cost)(params, cache=None)
+        assert res.shape == (3,)
+
+    @pytest.mark.parametrize(
+        "ret, mes",
+        [
+            ([qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))], "single return type"),
+            ([qml.state()], "Only Variance and Expectation"),
+        ],
+    )
+    def test_raises_for_jax_jit(self, execute_kwargs, interface, ret, mes):
+        """Tests multiple measurements and unsupported measurements raise an
+        error for the jit JAX interface."""
+        dev = qml.device("default.qubit", wires=2)
+        params = jnp.array([0.1, 0.2, 0.3])
+
+        def cost(a, cache):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.RY(a[2], wires=0)
+                [qml.apply(r) for r in ret]
+
+            res = qml.interfaces.batch.execute(
+                # Test only applicable for the jax jit interface
+                [tape],
+                dev,
+                cache=cache,
+                interface="jax-jit",
+                **execute_kwargs
+            )
+            return res[0][0]
+
+        with pytest.raises(ValueError, match=mes):

```

Comment:
Might be good to raise a custom exception here?

Reply:
What could be the goal there, to provide more specifics to the user? Would it entail creating a dedicated error for issues with the JIT interface?

Reply:
It's something I've been thinking about for a while; we rely too much on built-in Python exception classes in PL. We should instead be defining our own custom exceptions *when built-in exception classes do not make sense*, so that there is a level of control when performing `try`-`except`.

See the following:

- https://stackoverflow.com/questions/43772536/do-you-implement-your-own-exception-in-python
- https://stackoverflow.com/questions/57165380/when-should-i-declare-custom-exceptions

In this particular case, this does not feel like a `ValueError`, but a custom interface-specific error.

Reply:
Sounds good :+1: Added some custom error types in https://github.com/PennyLaneAI/pennylane/pull/2110/commits/a8c1b2c556b3dd88f47b18280e0e15e5b1e60739

Reply:
Thanks Antal! This looks great 😍

* Rather than `JAXMeasurementError`, which is very specific to the interface, I was thinking maybe we should course-grain it a bit more and just have `GradientUnsupportedError` error, that lives in a more general place? This way it could be used in a lot of different cases:
  
  * it could be used by any interface in any cases where a certain circuit operation or structure is not supported
  
  * It could be used by developers in other places in the stack. For example, the optimizers could do
  
    ```python
    try:
        qml.grad(circuit)(x)
    except GradientUnsupportedError as e:
        ...
    ```
  
    and know straight away that the users circuit is failing because the interface did not support an operation or measurement, rather than it being another issue or bug.
  
  * Plugins such as `lightning.qubit` could also raise the `GradientUnsupportedError` within their `adjoint` implementation, e.g., if `qml.state()` is returned :)

- I think `JAXForwardModeError` is also perhaps not essential; this is a case where it is better described by a `NotImplementedError` (but it definitely isn't a `ValueError`).

Reply:
Having `GradientUnsupportedError` would sound great, but I don't think creating a general error like that is within the scope of the PR here.

> I think JAXForwardModeError is also perhaps not essential; this is a case where it is better described by a NotImplementedError (but it definitely isn't a ValueError).

Hmm, frankly I'd think if we'd like to start using more specific errors such that `try`-`except` blocks can depend on the type of errors better, then a more special error than `NotImplementedError` would make sense here. Inheriting from that rather than `ValueError` might be a good middle ground.

Reply:
> Inheriting from that rather than ValueError might be a good middle ground.

Yes, good idea!

> Having `GradientUnsupportedError` would sound great, but I don't think creating a general error like that is within the scope of the PR here.

I agree that it is out of scope, although I find `JAXMeasurementError` too fine grained in terms of custom exceptions.

E.g., see the first link I posted above:

> Every exception class should cover a group of related uses (`ConfigError`, `BrowserError`, `DateParserError`). You definitely shouldn't create a new custom exception for every particular situation where something needs to be raised. That's what exception messages are for.

In this case, something like `InterfaceUnsupportedError` fits in that it covers a group of related uses (a common use case that is not supported by the interface), and the particular issue (that multiple measurements aren't supported with JAX) seems like a better fit for the error message.

Since this is out of scope, what do you think of the following?

* in this PR, just create and use `InterfaceUnsupportedError` directly in the JAX module?

* create an issue outlining a feature request to generalize this to all the interfaces, and also suggesting generalizations to gradient logic such as `GradientUnsupportedError`?

Reply:
Sure, added `InterfaceUnsupportedError` in https://github.com/PennyLaneAI/pennylane/pull/2110/commits/4f66be48a02a90fac4eff979cc68a71792b7bd6c, it inherits from `NotImplementedError`.

Also created https://github.com/PennyLaneAI/pennylane/issues/2127.